### ISSUE TITLE
Undefined variables are not reported when the target and value are of same name and an annotation is provided

### ISSUE THREAD

**Issue by Achxy** [microsoft/pyright#3852]
**Describe the bug**
No issues are raised by Pyright when the target and value of an assignment are of the same name and a type annotation is provided.
This does not happen if an annotation is not provided.

**To Reproduce**
Assign a variable to the same variable with any type annotation
Like:
```python
foo: int = foo
```
The above code is falsely deemed valid during typecheck but raises `NameError` during runtime.

**Expected behavior**
Pyright is expected to display errors when the value of an assignment is not declared.

**Medium**
VSCode extension, version: `v1.1.267`

**Additional context**
The above problem seems to be exclusive to assignments which contain a type annotation.
```python
# Start of the script, foo or bar is not defined as of now
foo: int = foo # This is valid (possible bug)
bar = bar # this is not valid (correct)
````

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.
